### ISSUE TITLE
Bug: "The above error…" should appear after the error message but appears before (for some errors)

### ISSUE THREAD

**Issue by sophiebits** [facebook/react#22656]
React version: both latest (17.0.2) and next (18.0.0-alpha-9c8161ba8-20211028)

## Steps To Reproduce

1. Render `<input>hello</input>` to trigger the "input is a void element tag" error

Link to code example: https://codesandbox.io/s/cocky-matan-ydmys

## The current behavior

Two log lines appear:
1. "The above error occurred in the \<input> component"
2. "input is a void element tag"

![image](https://user-images.githubusercontent.com/6820/139519976-0f3d55c3-638f-40b0-b09d-0df3ffdfa305.png)

## The expected behavior

They should be in the other order:

1. "input is a void element tag"
2. "The above error occurred in the \<input> component"

This problem doesn't occur in the same way if a component throws an error. Presumably this goes through a different error handling path because it comes from the host config.

If someone works on this, you may also want to review look at this issue at the same time:

* #18101

**Comment by A-Tavana** [comment]
خعاعاا

**Comment by aniler0** [comment]
The input tag should not have any children like this <input>Lorem ipsum dolor</input>. Instead of this you should use input and label tag if you want to write text out of input area.

**Comment by sophiebits** [comment]
Hi @aniler0 – I'm aware. :) But when someone makes this mistake, the reporting of the error should be a little different than it is today, and that's what this bug is about.

**Comment by sophiebits** [comment]
@Berci-Ionut I can't reproduce that. For me it does show the error:

![image](https://user-images.githubusercontent.com/6820/158668181-9ba6db6d-a8d9-4561-9541-80e8a5339e60.png)

If you can reproduce this in a standalone code sandbox, do feel free to link it here though.

**Comment by Berci-Ionut** [comment]
Hmm... Seems like is only reproducing when using NextJS [(example here)](https://codesandbox.io/s/elated-night-qpuyl2?file=/pages/index.js) . In this case I am not sure if the issue is related to react or rather NextJS

**Comment by sophiebits** [comment]
Still relevant (just verified in 18.0.0 stable) – this and #18101 are both interesting IMO.

**Comment by girishsontakke** [comment]
Hi @sophiebits, I would like to work on this issue.

**Comment by gaearon** [comment]
Sure, please feel free to. The first step would be to investigate why this is happening and post a comment here explaining the difference in behavior. Then we can look at the implementation.

**Comment by sylvesterAdewale** [comment]
The input tag should be a self closing tag

**Comment by gaearon** [comment]
@sylvesterAdewale This is not what the issue is about. The issue is about the ordering of the messages. @sophiebits already responded to this in https://github.com/facebook/react/issues/22656#issuecomment-956729030.

**Comment by MustafaEminn** [comment]
Oops. I have worked for a while on the issue. I guess I understand why messages are ordering like that. But I didn't know issues can be take.

**Comment by girishsontakke** [comment]
Hi @MustafaEminn, If you already have worked on this issue then you can surely write a patch for this, I will try to learn from your patch.

**Comment by MustafaEminn** [comment]
We can't run console methods after throwing the error. "The above error occurred..." error is logging by console.error method and the second error "... is a void element..." is throw a new error. If I want to tell in order.
assertValidProps method throw an error --> renderRootSync catch the error --> commitUpdateQueue call callCallback method and print the first error using console.error. Also, this callback set hasUncaughtError variable to true. --> commitRootImpl works again and throws an error which is "The above error occurred..." error.

What is the solution?
I guess changing the above word to below will be good. Because we can't run console methods after throwing the error. I already changed and tested with yarn test. There is no failure.

**Comment by gaearon** [comment]
I don’t think it sounds right to me that it’s not possible to fix this without changing the message. Why can’t we delay the console.error call until we know for sure the error has been logged? We could change where we log it from the updater callback to some other place.

**Comment by MustafaEminn** [comment]
We can change the throw statement position. 

We can throw the error here (here is logCapturedError method)
![image](https://user-images.githubusercontent.com/63938499/162626106-99e21b20-06f5-46e2-b46d-3585beb54aaf.png)

instead of here. I assume we had changed the throw statement to console.error.
![image](https://user-images.githubusercontent.com/63938499/162626043-1b065d80-f67f-4d75-9afc-4e04c14239c9.png)

Or another way to do it.
We can move the "The above error occurred" log into setTimeout (here is logCapturedError method).
![image](https://user-images.githubusercontent.com/63938499/162626380-5f70bc7a-480a-417b-ae87-8765f255bc5b.png)
Thus the log will be logged after the throw statement.

If we do even not catch the error, we should throw the error in logCapturedError because If we throw the error in assertValidProps method, execution will be interrupted and logCapturedError never going to work. That's why only one error could be logged.

**Comment by MustafaEminn** [comment]
We can delay the console.error with setTimeout but it does not seem an efficient way. If we log the error with console.error instead of throwing a new Error then we can change the order without setTimeout. Because we should log the first "is a void element" message and If we throw an error instantly (assume renderRootSync does not catch the error), execution will be interrupted. That's why I can't manage the second log order. Is interrupting the execution important? If we log the first error with console.error and the second error throw an error, handleError method catches the error and the second error won't be logged. 

There is good execution. If we change the code structure and log error with console.error instead of throwing a new Error, this doesn't look good. Maybe there is another way to do it but I'm new and I keep learning the React codebase 😊 

![image](https://user-images.githubusercontent.com/63938499/162634411-02408851-c884-498c-a372-61924e64fefe.png)

**Comment by abhijithp05** [comment]
Is this issue still open?? I would like to work on it

**Comment by sophiebits** [comment]
Still happens on next (18.3.0-next-4fcc9184a-20230217): https://codesandbox.io/s/stoic-hermann-x1kwd2

**Comment by VanshBajaj24** [comment]
> The input tag should be a self closing tag

Thanks a ton

**Comment by LeSingh1** [comment]
I checked this against current `main`, and the original ordering issue no longer reproduces.

For the original repro shape, rendering `<input>children</input>`, the observed error timeline is now:

1. The real error is reported first:
   `input is a void element tag...`
2. The React component addendum is reported after that:
   `An error occurred in the <input> component.`

From the current code path, this appears to have been fixed by the newer error reporting flow. `defaultOnUncaughtError` in `packages/react-reconciler/src/ReactFiberErrorLogger.js` calls `reportGlobalError(error)` first, then logs the React component addendum afterward.

I also wrote a focused local regression test for this ordering using the `<input>children</input>` case. The relevant test passes, and the broader related suites pass as well:

```bash
yarn test ReactDOMErrorReportingOrder
yarn test ReactDOMConsoleErrorReporting ReactDOMComponent ReactErrorBoundaries
```

Results:

```text
ReactDOMErrorReportingOrder: 1/1 passed
ReactDOMConsoleErrorReporting ReactDOMComponent ReactErrorBoundaries: 233/233 passed
```

Since the bug appears fixed on `main`, I have not opened a PR. Would maintainers want a small regression-test-only PR to lock in this ordering, or should this issue simply be closed as already fixed?
